merkle-maker/merkle_maker/publisher.py

```diff
@@ -0,0 +1,125 @@
+
+    @property
+    def deribit_fetcher(self) -> DeribitOptionsFetcher:
+        # We know for sure that fetchers[0] is DeribitOptionsFetcher, see assertions above.
+        return self.fetcher_client.fetchers[0]  # type: ignore[return-value]
+
+    async def publish_forever(self) -> Never:
+        """
+        Publish a new Merkle Feed on chain every [self.block_interval] block(s)
+        forever.
+        We store the merkle tree and the options used to generate the merkle root
+        to a Redis database that will get consumed by our Rust services.
+        """
+        current_block = await self.pragma_client.get_block_number()
+        while True:
+            logger.info(f"Current block: {current_block}")
+
+            if self._current_block_is_not_processed(current_block):
+                await self._publish_and_store(current_block=current_block)
+            else:
+                logger.info(f"🫷 Block {current_block} is already processed!\n")
+
+            next_block = current_block + self.block_interval
+            logger.info(f"⏳ Waiting for block {next_block}...")
+
+            while True:
+                await asyncio.sleep(self.time_to_wait_between_block_number_polling)
+                new_block = await self.pragma_client.get_block_number()
+                if new_block >= next_block:
+                    logger.info(f"⌛ ... reached block {new_block}!\n")
+                    current_block = new_block
+                    break
+
+    async def _publish_and_store(
+        self,
+        current_block: int,
+    ) -> None:
+        """
+        Retrieves the options from Deribit, publish the merkle root on-chain for
+        the network and block_number.
+        When done, store to Redis the merkle tree and the options used.
+        """
+        logger.info("🔍 Fetching the deribit options...")
+        entries = await self.fetcher_client.fetch()
+        logger.info("... fetched!")
+
+        logger.info("🎣 Publishing the merkle root onchain...")
+        await self.pragma_client.publish_entries(entries)  # type: ignore[arg-type]
+        logger.info("... published!")
+
+        logger.info("🏭 Storing the merkle tree & options in Redis...")
+        latest_data = self.deribit_fetcher.latest_data
+        success_store = self.redis_manager.store_block_data(
+            self.network, current_block, latest_data
+        )
+        if not success_store:
+            raise RuntimeError(f"Could not store data for block {current_block} to Redis.")
+        else:
+            logger.info("... stored!")
+
+        logger.info(f"✅ Block {current_block} done!\n")
+
+    def _current_block_is_not_processed(
+        self,
+        block_number: int,
+    ) -> bool:
+        """
+        Check if the current block is already processed, i.e if we already
+        published a merkle root onchain and stored the data into Redis.
+        """
```

Comment:
isn't it easier to just check what was published on-chain tho?

Reply:
wdym? I think it's harder to just look on-chain and associate a merkle root with a block number. We can probably do it by fetching the generic entry & checking the timestamp but checking Redis is easier/faster imo